client/client.go

```diff
@@ -122,25 +123,41 @@ func bind(cfg *Config, conn ldaputil.Connection) error {
 		return errors.New("unable to bind due to lack of configured password")
 	}
 
-	if cfg.BindDN == "" {
-		return errors.New("must provide binddn")
-	}
-
-	origErr := conn.Bind(cfg.BindDN, cfg.BindPassword)
-	if origErr == nil {
+	if cfg.UPNDomain != "" {
+		origErr := conn.Bind(fmt.Sprintf("%s@%s", ldaputil.EscapeLDAPValue(cfg.BindDN), cfg.UPNDomain), cfg.BindPassword)
+		if origErr == nil {
+			return nil
+		}
+		if !shouldTryLastPwd(cfg.LastBindPassword, cfg.LastBindPasswordRotation) {
+			return origErr
+		}
```

Comment:
I think errors should be wrapped and not skip levels when traversing between different systems. It makes it much easier to debug what exactly happened from a log file. I also think errors should never be swallowed, the code should either handle them or bubble the information up.

`return fmt.Errorf("unable to bind new password: %w", origErr)`

Reply:
I agree. I didn't modify this much to preserve the [logic](https://github.com/hashicorp/vault-plugin-secrets-ad/blob/main/plugin/client/client.go#L128-L156) from AD secrets, but I think it's a worthwhile improvement. I'm going to update this section of the code to incorporate your and Tom's comments. Will provide the commit for further feedback when done.